src/fsck.c

```diff
@@ -87,454 +91,341 @@ static int strtailcmp(const char *str, const char *tail) {
+	plaintext[0] &= 0x7f;
 
-			char *buf1 = malloc(len);
-			char *buf2 = malloc(len);
-			char *buf3 = malloc(len);
-
-			randomize(buf1, len);
-			buf1[0] &= 0x7f;
-			memset(buf2, 0, len);
-			memset(buf3, 0, len);
-			bool result = false;
-
-			if(rsa_public_encrypt(rsa_pub, buf1, len, buf2)) {
-				if(rsa_private_decrypt(rsa_priv, buf2, len, buf3)) {
-					if(memcmp(buf1, buf3, len)) {
-						result = true;
-					} else {
-						fprintf(stderr, "ERROR: public and private RSA keys do not match.\n");
-					}
-				} else {
-					fprintf(stderr, "ERROR: private RSA key does not work.\n");
-				}
+	if(rsa_public_encrypt(rsa_pub, plaintext, len, encrypted)) {
+		if(rsa_private_decrypt(rsa_priv, encrypted, len, decrypted)) {
+			if(memcmp(plaintext, decrypted, len) == 0) {
+				success = true;
 			} else {
-				fprintf(stderr, "ERROR: public RSA key does not work.\n");
+				fprintf(stderr, "ERROR: public and private RSA keys do not match.\n");
+				success = ask_fix_rsa_public_key(host_file, rsa_priv);
 			}
+		} else {
+			fprintf(stderr, "ERROR: private RSA key does not work.\n");
+			fprintf(stderr, "You can create a new RSA keypair with:\n\n");
+			print_tinc_cmd("generate-rsa-keys");
+		}
+	} else {
+		fprintf(stderr, "ERROR: public RSA key does not work.\n");
+		success = ask_fix_rsa_public_key(host_file, rsa_priv);
+	}
 
+	free(decrypted);
+	free(encrypted);
+	free(plaintext);
 
-			rsa_free(rsa_pub);
-			rsa_pub = NULL;
-
-			free(buf3);
-			free(buf2);
-			free(buf1);
-
-			if(!result) {
-				rsa_free(rsa_priv);
-				free(ecdsa_priv);
-				return 1;
-			}
-		}
+	return success;
+}
 
-		rsa_free(rsa_priv);
-		rsa_priv = NULL;
-	} else {
-		if(rsa_pub) {
-			fprintf(stderr, "WARNING: A public RSA key was found but no private key is known.\n");
-			rsa_free(rsa_pub);
-			rsa_pub = NULL;
-		}
+static bool check_rsa_pubkey(rsa_t *rsa_priv, rsa_t *rsa_pub, const char *host_file) {
+	if(!rsa_pub) {
+		fprintf(stderr, "WARNING: No (usable) public RSA key found.\n");
+		return ask_fix_rsa_public_key(host_file, rsa_priv);
 	}
 
-#endif
+	if(!rsa_priv) {
+		fprintf(stderr, "WARNING: A public RSA key was found but no private key is known.\n");
+		return true;
+	}
 
-	ecdsa_t *ecdsa_pub = NULL;
+	return test_rsa_keypair(rsa_priv, rsa_pub, host_file);
+}
+#endif // DISABLE_LEGACY
 
-	f = fopen(fname, "r");
+static bool test_ec_keypair(ecdsa_t *ec_priv, ecdsa_t *ec_pub, const char *host_file) {
+	// base64-encoded public key obtained from the PUBLIC key.
+	char *b64_pub_pub = ecdsa_get_base64_public_key(ec_pub);
 
-	if(f) {
-		ecdsa_pub = get_pubkey(f);
+	if(!b64_pub_pub) {
+		fprintf(stderr, "ERROR: public Ed25519 key does not work.\n");
+		return ask_fix_ec_public_key(host_file, ec_priv);
+	}
 
-		if(!ecdsa_pub) {
-			rewind(f);
-			ecdsa_pub = ecdsa_read_pem_public_key(f);
-		}
+	// base64-encoded public key obtained from the PRIVATE key.
+	char *b64_priv_pub = ecdsa_get_base64_public_key(ec_priv);
 
-		fclose(f);
+	if(!b64_priv_pub) {
```

Comment:
Shouldn't this test be done before checking `b64_pub_pub`? After all, we can regenerate the public key from the private key, but not the other way around.